### PR TITLE
Optimize SliverMainAxisGroup/SliverCrossAxisGroup paint function

### LINKED ISSUE #129214: SliverCrossAxisGroup & SliverMainAxisGroup should optimize painting
For both of these widgets, they do not currently consider if the child is visible before painting them. Skipping painting for children that are not visible would be a nice optimization, and would follow other widgets like RenderViewport.paint in the pattern of behavior.

```dart
// In paint
if (child.geometry!.visible) {
 paint...
}
```

cc @thkim1011

### PR DESCRIPTION
**Pull request by thkim1011**
This PR changes the paint functions for SliverMainAxisGroup and SliverCrossAxisGroup so that only visible slivers are painted.

Fixes https://github.com/flutter/flutter/issues/129214.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/test/widgets/sliver_cross_axis_group_test.dart [hunk 1]
```diff
@@ -806,6 +808,65 @@ void main() {
     // If renderHeader._lastStartedScrollDirection is not ScrollDirection.forward, then we shouldn't see the header at all.
     expect((renderHeader.parentData! as SliverPhysicalParentData).paintOffset.dy, equals(0.0));
   });
+
```
Comment (Piinks) [inline packages/flutter/test/widgets/sliver_cross_axis_group_test.dart:811]: Both the cross axis group and the main axis group tests here pass without the change to the paint method in the group widgets. They don't validate the change.

What they are checking is that the SliverToBoxAdapters are painting/not painting as expected, but not the sliver groups.

Comment (thkim1011) [reply packages/flutter/test/widgets/sliver_cross_axis_group_test.dart:811]: How would I solve this then? I'd need to verify that SliverCrossAxisGroup and SliverMainAxisGroup draws exactly two boxes, but I'm not sure how to do this.

Comment (Piinks) [reply packages/flutter/test/widgets/sliver_cross_axis_group_test.dart:811]: There are a couple different ways to do this. You can look up the Sliver*AxisGroup in the same way you are looking up the SliverToBoxAdapter.
You could mock out the class and increment a counter every time it paints.
There are probably a lot of different examples in existing tests if that helps. :)

File: packages/flutter/test/widgets/test_widgets.dart [hunk 2]
```diff
@@ -58,3 +59,33 @@ class FlipWidgetState extends State<FlipWidget> {
 void flipStatefulWidget(WidgetTester tester, { bool skipOffstage = true }) {
   tester.state<FlipWidgetState>(find.byType(FlipWidget, skipOffstage: skipOffstage)).flip();
 }
+
+// Test sliver which always attempts to paint itself whether it is visible or not.
+// Use for checking if slivers which take sliver children paints optimally.
+class RenderMockSliverToBoxAdapter extends RenderSliverToBoxAdapter {
```
Comment (Piinks) [inline packages/flutter/test/widgets/test_widgets.dart:65]: We were chatting about this yesterday, why not put this in the rendering directory under a file name like `sliver_utils.dart`? We typically try to keep utilities like this organized in that way. Importing the rendering library here also kind of indicates this should be in rendering.

**Review by Piinks — CHANGES_REQUESTED** [review]

File: packages/flutter/test/widgets/test_widgets.dart (hunk 2 above)
Comment (thkim1011) [reply packages/flutter/test/widgets/test_widgets.dart:65]: Ah I discovered the `test_widgets.dart` file after chatting but yeah it makes sense that this should go in rendering. I'll fix this.

**Review by Piinks — APPROVED** [review]
LGTM

**Pull request merged**